### DISCUSSION TITLE
Maturity of Bevy for serious 2D iOS projects?

### DISCUSSION THREAD

**Discussion by feelingsonice** [bevyengine/bevy#10351] [Q&A]
First of all, I understand these questions are difficult to give a straight forward answer to, but would love to get an estimate of my mileage if I were to use Bevy, because it does seems quite promising, so any input's appreciated.

I'm planning out a mobile game project that I intend to fully ship and maintain on the App Store. I previous used Godot for my projects but not a big fan of GDScript and I love Rust. Hence, I want to evaluate the feasibility of using Bevy for this project. Fair warning, I barely skimmed through the Bevy starter guide, so I'm a total beginner.

The game I plan to ship is a 2D iOS game in pixel arts. It is single player with in-app purchases but will frequently rely on server to client communications for things like weekly rewards, typical mobile game stuff. Here are some notable questions I had about Bevy:

* How does Bevy compare to Godot in the iOS plugin department? I intend to use a lot of the typical mobile functionalities like haptic feedback, screen swiping, notification sending, etc. Godot does have **some** iOS plugins for this but it's fairly limited. For Bevy, am I going to have to use FFI and bind to the iOS Swift or Objective C libraries? I've never done this before so it's daunting to me.
* How is Bevy's performance on iOS? Godot's mobile performance is suboptimal. A Reddit [post](https://www.reddit.com/r/godot/comments/17mojbv/why_is_godot_3ds_performance_so_bad_on_mobile/?utm_source=share&utm_medium=web2x&context=3) this morning talks about it. My game is 2D pixel art so it's not intensive but as I have learned with Godot, it's still fairly easy to blow up my phone with a simple game.
* I keep seeing conversation everywhere about how Bevy is missing important features, but it's difficult to get a grasp on what's missing, especially since my use case is fairly narrow (2D iOS pixel art). I don't care about editor support but would suck if I suddenly realize something absolutely critical isn't there a year into the project.

Again, highly appreciate anyone's input!

**Comment by alice-i-cecile** [comment]
@HackerFoo's [Noumenal](https://noumenal.app/) project is the most mature iOS project by quite a bit, but several people have successfully shipped iOS apps to the App Store.

>  I intend to use a lot of the typical mobile functionalities like haptic feedback, screen swiping, notification sending, etc. Godot does have some iOS plugins for this but it's fairly limited. For Bevy, am I going to have to use FFI and bind to the iOS Swift or Objective C libraries? I've never done this before so it's daunting to me.

I would expect a fair bit of FFI, but the community would likely be happy to collaborate on (and likely upstream) this work. But very little out-of-the-box at present.

> I keep seeing conversation everywhere about how Bevy is missing important features, but it's difficult to get a grasp on what's missing, especially since my use case is fairly narrow (2D iOS pixel art). I don't care about editor support but would suck if I suddenly realize something absolutely critical isn't there a year into the project.

I gave [a talk](https://docs.google.com/presentation/d/1_Lnw1p13fucGf41rcBvcXqlR7M2ADQxc6A1BUuB1FKo/edit?usp=sharing) about this in September 2023 :) For that particular project, you'll want to look at [seldom_pixel](https://github.com/Seldom-SE/seldom_pixel): it should give you an excellent starting experience. I would expect that limited engine audio support (use [bevy_kira_audio](https://github.com/NiklasEi/bevy_kira_audio)), and limited engine animation tools (use bevy_spine or one of the several sprite-based animation solutions) to be really the only missing features you'll run into. UI will be heavy on boilerplate but functional if you don't want complex widgets. Physics is fully third-party: bevy_xpbd and bevy_rapier are both defensible.

Overall: no serious blockers, other than maybe iOS specific missing features or performance that I'm not qualified to comment on.

**Reply by mockersf** [reply]
Specifically on iOS, what's missing is nicer integration with the platform. Things like running in the background, haptic feedback, ... But I would like to get those working within a year, and it will get even faster if you're around to test and contribute 🙂

**Reply by feelingsonice** [reply]
@alice-i-cecile and @mockersf thanks!

"You should use Bevy if: you're not actually making a game" sounds like a deal breaker for me... my goal is to make a polished and production-ready game and scale to as many users as possible, things like haptic feedbacks or running in the background sounds like obvious requirements.

Sounds like I'd be signing up for A LOT of work if I were to use Bevy?

**Reply by alice-i-cecile** [reply]
That quote from the slides is *one* of the reasons why you should. For your game, I don't really don't think a ton is missing. Just, like Francois said, platform integrations. Doing commercial dev in Bevy is likely always going to involve pretty close collaboration with the open source community around it: both the engine, and the plugins you rely on :)

I worked professionally on a pixel art platformer a year ago. There were very few missing pieces then, and things have only improved. Because you don't need skeletal animation, an editor, advanced spatial audio or AAA PBR graphics, you can bypass a lot of the missing features.

**Reply by feelingsonice** [reply]
@alice-i-cecile would you mind elaborating a bit more on your point from the talk:

>Your game world doesn’t rely on hand-authored levels.

I envision my game to have **some** hand-made levels that'll be recycled through content shuffling and randomization. I wouldn't be creating thousands of candy crush like levels but there are multiple maps to be made (kind of like the rooms in Hades if you ever played it).

**Reply by alice-i-cecile** [reply]
Yeah, so that's not too bad :) For 2D, I'd probably recommend using ldtk or Tiled as a level editor, then using one of the existing tilemap crates with support for them.

In 3D, I think my preference is probably to use [Blender Bevy Components](https://github.com/kaosat-dev/Blender_bevy_components_workflow), and use Blender itself as your editor.

For a small team, both of these approaches are quite good! The challenge only really starts to come when you want sprawling levels, with a team of less-technical designers, scripted events and so on.

**Reply by mrgzi** [reply]
Those who want to watch Alice's seminar can do so using [this link](https://www.youtube.com/watch?v=VpiprNBEZsk&ab_channel=GOSIMFoundation) :)

**Reply by alice-i-cecile** [reply]
That's actually the other talk I gave unfortunately! The talk linked should be up on that YouTube channel in the next week or so though.

**Comment by mockersf** [comment]
> I intend to use a lot of the typical mobile functionalities like haptic feedback, screen swiping, notification sending, etc. Godot does have some iOS plugins for this but it's fairly limited. For Bevy, am I going to have to use FFI and bind to the iOS Swift or Objective C libraries? I've never done this before so it's daunting to me.

Haptic feedback and gesture (screen swiping) should have direct support in Bevy. I'm trying to improve gestures currently (https://github.com/rust-windowing/winit/pull/3130) so that should come soon~ish

Notifications, in app purchases and maybe other APIs on the other hands, will probably stay in third party plugins. There are a few people on Discord getting support for ads and in app purchase, so it's definitely possible.

From my experiments, the way integration with iOS native languages are:
- you expose rust code through the C FFI and call it in Swift, that's fairly easy
- you expose native API through the Objective C version and call it in Rust, that's a bit boilerplaty but OK to do. iOS API have good documentation

In both cases, you should be able to find help!

**Reply by feelingsonice** [reply]
>you expose native API through the Objective C version and call it in Rust, that's a bit boilerplaty but OK to do. iOS API have good documentation.

That sounds like the better approach. Are there boilerplates for this already? If not I'd imagine it to be a lot harder.

**Reply by mockersf** [reply]
you can look at the [`objc2` crate](https://github.com/madsmtm/objc2) and maybe [`icrate`](https://docs.rs/icrate/latest/icrate/), and how `winit` does it ([binding to `touch` for example](https://github.com/rust-windowing/winit/blob/master/src/platform_impl/ios/uikit/touch.rs))

**Comment by mockersf** [comment]
> How is Bevy's performance on iOS? Godot's mobile performance is suboptimal. A Reddit [post](https://www.reddit.com/r/godot/comments/17mojbv/why_is_godot_3ds_performance_so_bad_on_mobile/?utm_source=share&utm_medium=web2x&context=3) this morning talks about it. My game is 2D pixel art so it's not intensive but as I have learned with Godot, it's still fairly easy to blow up my phone with a simple game.

I'm using Bevy on an iPhone 13 and an iPad Pro M2. Not the oldest devices, but it works without issues on both. And I haven't heard issues about perf on iOS
